pandas/tests/groupby/test_groupby_dropna.py

```diff
@@ -276,3 +276,60 @@ def test_groupby_dropna_datetime_like_data(
     expected = pd.DataFrame({"values": values}, index=pd.Index(indexes, name="dt"))
 
     tm.assert_frame_equal(grouped, expected)
+
+
+@pytest.mark.parametrize(
+    "dropna, df_cols_in, df_cols_out, levels",
+    [
+        pytest.param(
+            False,
+            {"groups": ["a", "a", "b", np.nan], "values": [10, 10, 20, 30]},
+            {"values": [0, 1, 0, 0]},
+            ["a", "b", np.nan],
+            id="dropna_false_has_nan",
+        ),
+        pytest.param(
+            True,
+            {"groups": ["a", "a", "b", np.nan], "values": [10, 10, 20, 30]},
+            {"values": [0, 1, 0]},
+            None,
+            id="dropna_true_has_nan",
+        ),
+        pytest.param(
+            # no nan in "groups"; dropna=True|False should be same.
+            False,
+            {"groups": ["a", "a", "b", "c"], "values": [10, 10, 20, 30]},
+            {"values": [0, 1, 0, 0]},
+            None,
+            id="dropna_false_no_nan",
+        ),
+        pytest.param(
+            # no nan in "groups"; dropna=True|False should be same.
+            True,
+            {"groups": ["a", "a", "b", "c"], "values": [10, 10, 20, 30]},
+            {"values": [0, 1, 0, 0]},
+            None,
+            id="dropna_true_no_nan",
+        ),
+    ],
+)
+def test_groupby_dropna_multi_index_dataframe_apply(
+    dropna, df_cols_in, df_cols_out, levels
+):
+    # GH 35889
+    # `groupby` with `dropna=False` and `apply` returning DataFrame of different
+    # sizes raises error if grouped column has nan values.
+
+    df = pd.DataFrame(df_cols_in)
+    dfg = df.groupby("groups", dropna=dropna)
+    rv = dfg.apply(lambda grp: pd.DataFrame({"values": list(range(len(grp)))}))
```

Comment:
another nitpick: call this `result`

Reply:
also you can feed `range` directly to the DataFrame constructor, no need to convert to list